tests/unit_tests.cpp

```diff
@@ -0,0 +1,245 @@
+              if(lid < i) scratch[lid] += scratch[lid + i];
+            });
+          }
+          wg.parallel_for_work_item([&](cl::sycl::h_item<1> item) {
+            const size_t lid = item.get_local_id()[0];
+            if(lid == 0) acc[item.get_global_id()] = scratch[0];
+          });
+        });
+    });
+  }
+
+  for(size_t i = 0; i < global_size / local_size; ++i) {
+    size_t expected = 0;
+    for(size_t j = 0; j < local_size; ++j) expected += i * local_size + j;
+    size_t computed = host_buf[i * local_size];
+    BOOST_TEST(computed == expected);
+  }
+}
+
+BOOST_AUTO_TEST_CASE(local_memory) {
+  constexpr size_t local_size = 256;
+  constexpr size_t global_size = 1024;
+
+  cl::sycl::queue queue;
+  std::vector<int> host_buf;
+  for(size_t i = 0; i < global_size; ++i) {
+    host_buf.push_back(static_cast<int>(i));
+  }
+
+  {
+    cl::sycl::buffer<int, 1> buf{host_buf.data(), host_buf.size()};
+    queue.submit([&](cl::sycl::handler& cgh) {
+      using namespace cl::sycl::access;
+      auto acc = buf.get_access<mode::read_write>(cgh);
+      auto scratch = cl::sycl::accessor<int, 1, mode::read_write, target::local>
+        {local_size, cgh};
+
+      cgh.parallel_for<class local_memory_reduction>(
+        cl::sycl::nd_range<1>{global_size, local_size},
+        [=](cl::sycl::nd_item<1> item) {
+          const auto lid = item.get_local(0);
+          scratch[lid] = acc[item.get_global()];
+          item.barrier();
+          for(size_t i = local_size/2; i > 0; i /= 2) {
+            if(lid < i) scratch[lid] += scratch[lid + i];
+            item.barrier();
+          }
+          if(lid == 0) acc[item.get_global()] = scratch[0];
+        });
+    });
+  }
+
+  for(size_t i = 0; i < global_size / local_size; ++i) {
+    size_t expected = 0;
+    for(size_t j = 0; j < local_size; ++j) expected += i * local_size + j;
+    size_t computed = host_buf[i * local_size];
+    BOOST_TEST(computed == expected);
+  }
+}
+
+BOOST_AUTO_TEST_CASE(placeholder_accessors) {
+  using namespace cl::sycl::access;
+  constexpr size_t num_elements = 4096 * 1024;
+
+  cl::sycl::queue queue;
+  cl::sycl::buffer<int, 1> buf{num_elements};
+
+  {
+    auto acc = buf.get_access<mode::discard_write>();
+    for(size_t i = 0; i < num_elements; ++i) acc[i] = static_cast<int>(i);
+  }
+
+  cl::sycl::accessor<int, 1, mode::read_write, target::global_buffer, placeholder::true_t>
+    ph_acc{buf};
+
+  queue.submit([&](cl::sycl::handler& cgh) {
+    auto ph_acc_copy = ph_acc; // Test that placeholder accessors can be copied
+    cgh.require(ph_acc_copy);
+    cgh.parallel_for<class placeholder_accessors>(cl::sycl::range<1>{num_elements},
+      [=](cl::sycl::id<1> tid) {
+        ph_acc_copy[tid] *= 2;
+      });
+  });
+
+  {
+    auto acc = buf.get_access<mode::read>();
+    for(size_t i = 0; i < num_elements; ++i) {
+      BOOST_REQUIRE(acc[i] == 2 * i);
+    }
+  }
+}
+
+BOOST_AUTO_TEST_CASE(multiple_queues) {
```

Comment:
This test actually tests automatic synchonization and ordering of concurrent operations in the task graph, the multiple queues are more of an implementation detail to obtain concurrent operations on NVIDIA GPUs.. Queues are very simple objects with barely any logic from hipSYCL's point of view, so this test should perhaps be named `task_graph_synchronization` or something like that?

Maybe it's also a good idea to execute this test in a loop since synchronization issues tend to only pop up sometimes...